File: DSA/linkedlist/Linkedlist.cpp
#include <iostream>
using namespace std;

// Node class
class Node {
public:
    int data;
    Node* next;
    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};

// Linked list class
class LinkedList {
private:
    Node* head;
public:
    LinkedList() {
        head = NULL;
    }

    // Add node to the end of the list
    void add(int data) {
        Node* newNode = new Node(data);
        if (head == NULL) {
            head = newNode;
            return;
        }
        Node* current = head;
        while (current->next != NULL) {
            current = current->next;
        }
        current->next = newNode;
    }

    // Delete node at given index
    void remove(int index) {
        if (head == NULL) {
            cout << "List is empty" << endl;
            return;
        }
        Node* current = head;
        if (index == 0) {
            head = current->next;
            delete current;
            return;
        }
        for (int i = 0; current != NULL && i < index-1; i++) {
            current = current->next;
        }
        if (current == NULL || current->next == NULL) {
            cout << "Index out of range" << endl;
            return;
        }
        Node* next = current->next->next;
        delete current->next;
        current->next = next;
    }

    // Update value of node at given index
    void update(int index, int data) {
        if (head == NULL) {
            cout << "List is empty" << endl;
            return;
        }
        Node* current = head;
        for (int i = 0; current != NULL && i < index; i++) {
            current = current->next;
        }
        if (current == NULL) {
            cout << "Index out of range" << endl;
            return;
        }
        current->data = data;
    }

    // Sort the linked list in ascending order
    void sort() {
        if (head == NULL) {
            return;
        }
        bool swapped;
        Node* ptr1;
        Node* lptr = NULL;
        do {
            swapped = false;
            ptr1 = head;
            while (ptr1->next != lptr) {
                if (ptr1->data > ptr1->next->data) {
                    swap(ptr1->data, ptr1->next->data);
                    swapped = true;
                }
                ptr1 = ptr1->next;
            }
            lptr = ptr1;
        } while (swapped);
    }

    // Search for node with given data and return its index
    int search(int data) {
        if (head == NULL) {
            cout << "List is empty" << endl;
            return -1;
        }
        Node* current = head;
        int index = 0;
        while (current != NULL) {
            if (current->data == data) {
                return index;
            }
            current = current->next;
            index++;
        }
        cout << "Data not found" << endl;
        return -1;
    }

    // Print the linked list
    void print() {
        Node* current = head;
        while (current != NULL) {
            cout << current->data << " ";
            current = current->next;
        }
        cout << endl;
    }
};

int main() {
    LinkedList list;
    int choice, data, index;

    do {
        cout << "\nChoose an option:\n";
        cout << "1. Add a node\n";
        cout << "2. Remove a node\n";
        cout << "3. Update a node\n";
        cout << "4. Sort the list\n";
        cout << "5. Search for a node\n";
        cout << "6. Print the list\n";
        cout << "7. Exit\n";
        cin >> choice;

        switch (choice) {
            case 1:
                cout << "Enter data to add: ";
                cin >> data;
                list.add(data);
                break;
            case 2:
                cout << "Enter index to remove: ";
                cin >> index;
                list.remove(index);
                break;
            case 3:
                cout << "Enter index to update: ";
                cin >> index;
                cout << "Enter new data: ";
                cin >> data;
                list.update(index, data);
                break;
            case 4:
                list.sort();
                cout << "List sorted in ascending order" << endl;
                break;
            case 5:
                cout << "Enter data to search: ";
                cin >> data;
                index = list.search(data);
                if (index != -1) {
                    cout << "Data found at index " << index << endl;
                }
                break;
            case 6:
                list.print();
                break;
            case 7:
                cout << "Exiting program" << endl;
                break;
            default:
                cout << "Invalid choice, please try again" << endl;
                break;
        }
    } while (choice != 7);

    return 0;
}

File: DSA/LinkedList/LinkedList.java
import java.util.Scanner;
import java.util.InputMismatchException;

class CustomLinkedList {
  int data;
  CustomLinkedList next;
   CustomLinkedList start = null;

  int countCustomLinkedList() //count nodes
  {
    int c = 0;
    CustomLinkedList tempNode = start;
    while (tempNode != null) {
      c++;
      tempNode = tempNode.next;
    }
    return c;
  }

  CustomLinkedList msort(CustomLinkedList first,CustomLinkedList second) //merge sort to merge two sorted LinkedList
  {
    CustomLinkedList result;
    if (first == null)
      return second;
    if (second == null)
      return first;
    if (first.data <= second.data) {
      result = first;
      result.next = msort(first.next, second);
    } else {
      result = second;
      result.next = msort(first, second.next);
    }
    return result;
  }
  
  void sort() // main method to sort
  {
    start = split(start);
    showAll();
  }
  
  CustomLinkedList split(CustomLinkedList tempNode) // to break a CustomLinkedList into two parts
  {
    if (tempNode == null || tempNode.next == null) {
      return tempNode;
    }

    CustomLinkedList first = getMiddle(tempNode); // to find middle
    CustomLinkedList second = first.next;

    first.next = null;

    CustomLinkedList left = split(tempNode);
    CustomLinkedList right = split(second);

    CustomLinkedList sortedlist = msort(left, right);
    return sortedlist;
  }

  public static CustomLinkedList getMiddle(CustomLinkedList head) //method to find middle
  {
    if (head == null)
      return head;

    CustomLinkedList slow = head, fast = head;

    while (fast.next != null && fast.next.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }
    return slow;
  }

  void update(int pre, int nw) // method to update a CustomLinkedList
  {
    CustomLinkedList tempNode = start;
    while (tempNode != null && tempNode.data != pre) {
      tempNode = tempNode.next;
    }
    if (tempNode == null)
      System.out.println("Data not found");
    else
      tempNode.data = nw;
  }

  void showAll() //method to print 
  {
    CustomLinkedList tempNode = start;
    while (tempNode != null) {
      System.out.print(tempNode.data + "  ");
      tempNode = tempNode.next;
    }
    System.out.println();
  }

  void deleteAtB() //deletion at beginning
  {
    if (start == null) {
      System.out.println("list is empty");
      return;
    }
    CustomLinkedList tempNode1 = start;
    start = start.next;
    tempNode1 = null;
  }

  void deleteAtE() // deletion at ending
  {
    if (start == null || start.next == null) {
      deleteAtB();
      return;
    }
    CustomLinkedList tempNode1 = start;
    CustomLinkedList tempNode2 = start;
    while (tempNode1.next != null) {
      tempNode2 = tempNode1;
      tempNode1 = tempNode1.next;
    }
    tempNode2.next = null;
    tempNode1 = null;
  }

  void deleteAtSP(int position) // deletion at specific point
  {
    if (position <= 1) {
      deleteAtB();
      return;
    }
    if (position > countCustomLinkedList()) {
      deleteAtE();
      return;
    }
    CustomLinkedList tempNode1 = start;
    CustomLinkedList t;
    int index;
    for (index = 1; index < position - 1; index++) {
      tempNode1 = tempNode1.next;
    }
    t = tempNode1.next;
    tempNode1.next = t.next;
    t = null;
  }

  void insertAtB(int data) // insertion at beginning
  {
    CustomLinkedList tempNode = new CustomLinkedList();
    tempNode.data = data;
    tempNode.next = start;
    start = tempNode;
  }

  void insertAtE(int data) // insertion at ending
  {
    if (start == null) {
      insertAtB(data);
      return;
    }
    CustomLinkedList tempNode = new CustomLinkedList();
    CustomLinkedList tempNode1 = start;
    tempNode.data = data;
    tempNode.next = null;
    while (tempNode1.next != null)
      tempNode1 = tempNode1.next;

    tempNode1.next = tempNode;
  }

  boolean search(int data) // searching an element 
  {
    CustomLinkedList tempNode = start;
    while (tempNode != null) {
      if (tempNode.data == data) return true;
      tempNode = tempNode.next;
    }
    return false;
  }

  void insertAtSP(int data, int position) // insertion at specific point
  {
    if (position <= 1) {
      insertAtB(data);
      return;
    }
    if (position >= countCustomLinkedList()) {
      insertAtE(data);
      return;
    }
    CustomLinkedList tempNode = new CustomLinkedList();
    CustomLinkedList tempNode1 = start;
    tempNode.data = data;
    for (int index = 1; index < position - 1; index++)
      tempNode1 = tempNode1.next;

    tempNode.next = tempNode1.next;
    tempNode1.next = tempNode;
  }

}
class Main {
  public static void main(String[] args) {
    CustomLinkedList start = new CustomLinkedList();
    int value1, value2;
    while (true) {
      System.out.println("===========================");
      System.out.println("*****Enter your choice*****");
      System.out.println("1.Insert at beginning");
      System.out.println("2.Insert at ending");
      System.out.println("3.Insert at specific position");
      System.out.println("4.Deletion at beginning");
      System.out.println("5.Deletion at ending");
      System.out.println("6.Deletion at specific position");
      System.out.println("7.Update");
      System.out.println("8.Search");
      System.out.println("9.ShowAll");
      System.out.println("10. Sort");
      System.out.println("11.Exit");
      System.out.println("===========================");
      try {
        Scanner sc = new Scanner(System.in);
        int condition = sc.nextInt();
        switch (condition) {
        case 1:
          System.out.println("Enter a number ");
          value1 = sc.nextInt();
          start.insertAtB(value1);
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 2:
          System.out.println("Enter a number ");
          value2 = sc.nextInt();
          start.insertAtE(value2);
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 3:
          System.out.println("Enter a number ");
          value1 = sc.nextInt();
          System.out.println("Enter a position ");
          value2 = sc.nextInt();
          start.insertAtSP(value1, value2);
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 4:
          start.deleteAtB();
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 5:
          start.deleteAtE();
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 6:
          System.out.println("Enter a position ");
          value2 = sc.nextInt();
          start.deleteAtSP(value2);
          System.out.println("Done \nList is :");
		  start.showAll();
          break;

        case 7:
          System.out.println("Enter no. to be replaced ");
          value1 = sc.nextInt();
          System.out.println("Enter new number ");
          value2 = sc.nextInt();
          start.update(value1, value2);
          System.out.println("List is :");
		  start.showAll();
          break;

        case 8:
          System.out.println("Enter a number ");
          value2 = sc.nextInt();
          System.out.println(start.search(value2));
          break;

        case 9:
          start.showAll();
          break;

        case 10:
          start.sort();
          break;

        case 11:
          System.exit(0);
          break;

        default:
          System.out.println("Invalid");
          break;
        }
      } catch (InputMismatchException e) {
        System.out.println("Enter valid option");
      }
    }
  }
}